File: Tellma/Entities/AdminUser.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Tellma.Entities
{
    [StrongEntity]
    [EntityDisplay(Singular = "AdminUser", Plural = "AdminUsers")]
    public class AdminUserForSave<TPermission> : EntityWithKey<int>
    {
        [Display(Name = "Name")]
        [Required]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name { get; set; }

        [Display(Name = "User_Email")]
        [Required]
        [NotNull]
        [EmailAddress]
        [StringLength(255)]
        public string Email { get; set; }

        [Display(Name = "User_Permissions")]
        [ForeignKey(nameof(AdminPermission.AdminUserId))]
        public List<TPermission> Permissions { get; set; }
    }

    public class AdminUserForSave : AdminUserForSave<AdminPermissionForSave>
    {

    }

    public class AdminUser : AdminUserForSave<AdminPermission>
    {
        public string ExternalId { get; set; }

        [Display(Name = "State")]
        [NotNull]
        [ChoiceList(new object[] { "Invited", "Member" },
            new string[] { "User_Invited", "User_Member" })]
        public string State { get; set; }

        [Display(Name = "User_LastActivity")]
        public DateTimeOffset? LastAccess { get; set; }

        [Display(Name = "IsActive")]
        [NotNull]
        [AlwaysAccessible]
        public bool? IsActive { get; set; }

        [Display(Name = "CreatedAt")]
        [NotNull]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        [NotNull]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        [NotNull]
        public DateTimeOffset? ModifiedAt { get; set; }

        [Display(Name = "ModifiedBy")]
        [NotNull]
        public int? ModifiedById { get; set; }

        // For Query

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public AdminUser CreatedBy { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(ModifiedById))]
        public AdminUser ModifiedBy { get; set; }
    }
}

File: Tellma/ClientApp/src/app/data/entities/admin-user.ts
// tslint:disable:variable-name
import { EntityDescriptor } from './base/metadata';
import { WorkspaceService } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { EntityForSave } from './base/entity-for-save';
import { AdminSettingsForClient } from '../dto/admin-settings-for-client';
import { AdminPermissionForSave, AdminPermission } from './admin-permission';

export interface AdminUserForSave<TPermission = AdminPermissionForSave> extends EntityForSave {
    Name?: string;
    Email?: string;
    Permissions?: TPermission[];
}

export interface AdminUser extends AdminUserForSave<AdminPermission> {
    State: string;
    IsActive?: boolean;
    ExternalId?: string;
    LastAccess?: string;
    CreatedAt?: string;
    CreatedById?: number | string;
    ModifiedAt?: string;
    ModifiedById?: number | string;
}

let _settings: AdminSettingsForClient;
let _cache: EntityDescriptor;

export function metadata_AdminUser(wss: WorkspaceService, trx: TranslateService): EntityDescriptor {
    const ws = wss.admin;
    // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
    if (ws.settings !== _settings) {
        _settings = ws.settings;
        _cache = {
            collection: 'AdminUser',
            titleSingular: () => trx.instant('AdminUser'),
            titlePlural: () => trx.instant('AdminUsers'),
            select: ['Name'],
            apiEndpoint: 'admin-users',
            masterScreenUrl: 'admin-users',
            orderby: () => ['Name'],
            inactiveFilter: 'IsActive eq true',
            format: (item: AdminUserForSave) => item.Name,
            isAdmin: true,
            properties: {
                Id: { datatype: 'numeric', control: 'number', label: () => trx.instant('Id'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
                Name: { datatype: 'string', control: 'text', label: () => trx.instant('Name') },
                Email: { datatype: 'string', control: 'text', label: () => trx.instant('User_Email') },
                State: {
                    datatype: 'string',
                    control: 'choice',
                    label: () => trx.instant('State'),
                    choices: ['Invited', 'Member'],
                    format: (c: string) => {
                        switch (c) {
                            case 'Invited': return trx.instant('User_Invited');
                            case 'Member': return trx.instant('User_Member');
                            default: return c;
                        }
                    },
                    color: (c: string) => {
                        switch (c) {
                            case 'Invited': return '#6c757d';
                            case 'Member': return '#28a745';
                            default: return c;
                        }
                    }
                },
                LastAccess: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('User_LastActivity') },
                IsActive: { datatype: 'bit', control: 'check', label: () => trx.instant('IsActive') },
                CreatedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('CreatedAt') },
                CreatedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('CreatedBy'), foreignKeyName: 'CreatedById' },
                ModifiedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('ModifiedAt') },
                ModifiedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('ModifiedBy'), foreignKeyName: 'ModifiedById' }
            }
        };
    }

    return _cache;
}
